File: src/CobolDocumentor/Export/GraphifyExporter.cs
using System.Text.Json;
using System.Text.RegularExpressions;
using CobolDocumentor.Model;

namespace CobolDocumentor.Export;

/// <summary>Exports parsed COBOL programs to Graphify's simple nodes/edges shape.</summary>
public static class GraphifyExporter
{
    /// <summary>Builds a graph from a parsed COBOL program.</summary>
    public static GraphifyGraph Export(CobolProgramModel program)
    {
        var graph = new GraphifyGraph();
        var seenNodes = new HashSet<string>(StringComparer.OrdinalIgnoreCase);
        var seenEdges = new HashSet<string>(StringComparer.OrdinalIgnoreCase);
        var programId = Id("program", program.Name);

        AddNode(graph, seenNodes, programId, program.Name, "program", new Dictionary<string, object?> { ["source_file"] = program.SourceFile });

        foreach (var variable in program.MemoryStack.Variables)
        {
            var variableId = ExportVariable(graph, seenNodes, seenEdges, variable);
            AddEdge(graph, seenEdges, programId, variableId, "declares");
        }

        for (var i = 0; i < program.Paragraphs.Count; i++)
        {
            var paragraph = program.Paragraphs[i];
            var paragraphId = Id("paragraph", $"{program.Name}:{paragraph.Name}:{i}");
            AddNode(graph, seenNodes, paragraphId, paragraph.Name, "paragraph", new Dictionary<string, object?> { ["source_file"] = program.SourceFile });
            AddEdge(graph, seenEdges, programId, paragraphId, "contains");

            for (var j = 0; j < paragraph.Statements.Count; j++)
            {
                ExportStatement(graph, seenNodes, seenEdges, paragraphId, paragraph.Statements[j], j, program.MemoryStack.KnownVariables);
            }
        }

        return graph;
    }

    /// <summary>Writes a graph export to a JSON file.</summary>
    public static void Write(CobolProgramModel program, string outputPath)
    {
        var graph = Export(program);
        var directory = Path.GetDirectoryName(outputPath);
        if (!string.IsNullOrWhiteSpace(directory))
        {
            Directory.CreateDirectory(directory);
        }

        File.WriteAllText(outputPath, JsonSerializer.Serialize(graph, new JsonSerializerOptions { WriteIndented = true }));
    }

    private static string ExportVariable(GraphifyGraph graph, HashSet<string> seenNodes, HashSet<string> seenEdges, CobolVariable variable)
    {
        var variableId = Id("variable", variable.Name);
        AddNode(graph, seenNodes, variableId, variable.Name, "variable", new Dictionary<string, object?>
        {
            ["level"] = variable.Level,
            ["pic"] = variable.Picture,
            ["usage"] = variable.Usage,
            ["value"] = variable.Value,
            ["redefines"] = variable.Redefines,
            ["occurs"] = variable.Occurs,
            ["occurs_min"] = variable.OccursMin,
            ["occurs_max"] = variable.OccursMax,
            ["depends_on"] = variable.DependsOn,
            ["is_group"] = variable.IsGroup
        });

        if (!string.IsNullOrWhiteSpace(variable.Redefines))
        {
            var targetId = Id("variable", variable.Redefines);
            AddNode(graph, seenNodes, targetId, variable.Redefines, "variable", new Dictionary<string, object?> { ["inferred"] = true });
            AddEdge(graph, seenEdges, variableId, targetId, "redefines");
        }

        if (!string.IsNullOrWhiteSpace(variable.DependsOn))
        {
            var targetId = Id("variable", variable.DependsOn);
            AddNode(graph, seenNodes, targetId, variable.DependsOn, "variable", new Dictionary<string, object?> { ["inferred"] = true });
            AddEdge(graph, seenEdges, variableId, targetId, "occurs_depends_on");
        }

        foreach (var condition in variable.Conditions)
        {
            var conditionId = ExportVariable(graph, seenNodes, seenEdges, condition);
            AddEdge(graph, seenEdges, variableId, conditionId, "defines_condition");
        }

        foreach (var child in variable.Children)
        {
            var childId = ExportVariable(graph, seenNodes, seenEdges, child);
            AddEdge(graph, seenEdges, variableId, childId, "contains");
        }

        return variableId;
    }

    private static void ExportStatement(GraphifyGraph graph, HashSet<string> seenNodes, HashSet<string> seenEdges, string parentId, CobolStatement statement, int index, IReadOnlySet<string> knownVariables)
    {
        var statementId = Id("statement", $"{parentId}:{index}:{statement.RawText}");
        AddNode(graph, seenNodes, statementId, Short(statement.RawText), "statement", new Dictionary<string, object?> { ["statement_type"] = statement.Kind.ToString().ToLowerInvariant() });
        AddEdge(graph, seenEdges, parentId, statementId, "contains");
        AddSemanticEdges(graph, seenNodes, seenEdges, statementId, statement);

        foreach (var variableName in ReferencedVariables(statement.RawText, knownVariables))
        {
            var variableId = Id("variable", variableName);
            AddNode(graph, seenNodes, variableId, variableName, "variable", new Dictionary<string, object?> { ["inferred"] = true });
            AddEdge(graph, seenEdges, statementId, variableId, "references");
        }

        for (var i = 0; i < statement.Children.Count; i++)
        {
            ExportStatement(graph, seenNodes, seenEdges, statementId, statement.Children[i], i, knownVariables);
        }

        for (var i = 0; i < statement.ElseChildren.Count; i++)
        {
            ExportStatement(graph, seenNodes, seenEdges, statementId, statement.ElseChildren[i], i, knownVariables);
        }
    }

    private static void AddSemanticEdges(GraphifyGraph graph, HashSet<string> seenNodes, HashSet<string> seenEdges, string statementId, CobolStatement statement)
    {
        var raw = statement.RawText.Trim();
        var upper = raw.ToUpperInvariant();
        if (upper.StartsWith("CALL ", StringComparison.Ordinal))
        {
            var tokens = SplitTokens(raw);
            var target = statement.IsZCallPgm && tokens.Count >= 4 ? tokens[3] : tokens.ElementAtOrDefault(1);
            if (!string.IsNullOrWhiteSpace(target))
            {
                var programId = Id("program", CleanToken(target));
                AddNode(graph, seenNodes, programId, CleanToken(target), "program", new Dictionary<string, object?> { ["inferred"] = true });
                AddEdge(graph, seenEdges, statementId, programId, "calls");
            }

            if (statement.IsZCallPgm)
            {
                foreach (var copyArgument in tokens.Skip(4))
                {
                    var copyId = Id("copy", CleanToken(copyArgument));
                    AddNode(graph, seenNodes, copyId, CleanToken(copyArgument), "copy", new Dictionary<string, object?> { ["inferred"] = true });
                    AddEdge(graph, seenEdges, statementId, copyId, "uses_copy_argument");
                }
            }
        }
        else if (upper.StartsWith("PERFORM ", StringComparison.Ordinal))
        {
            foreach (var target in PerformTargets(raw))
            {
                var paragraphId = Id("paragraph", target);
                AddNode(graph, seenNodes, paragraphId, target, "paragraph", new Dictionary<string, object?> { ["inferred"] = true });
                AddEdge(graph, seenEdges, statementId, paragraphId, "performs");
            }
        }
        else if (upper.StartsWith("READ ", StringComparison.Ordinal))
        {
            AddFileEdge(graph, seenNodes, seenEdges, statementId, raw, "reads_from");
        }
        else if (upper.StartsWith("WRITE ", StringComparison.Ordinal) || upper.StartsWith("REWRITE ", StringComparison.Ordinal))
        {
            AddFileEdge(graph, seenNodes, seenEdges, statementId, raw, "writes_to");
        }
    }

    private static void AddFileEdge(GraphifyGraph graph, HashSet<string> seenNodes, HashSet<string> seenEdges, string statementId, string raw, string relation)
    {
        var target = SplitTokens(raw).ElementAtOrDefault(1);
        if (string.IsNullOrWhiteSpace(target))
        {
            return;
        }

        var fileId = Id("file", target);
        AddNode(graph, seenNodes, fileId, target, "file", new Dictionary<string, object?> { ["inferred"] = true });
        AddEdge(graph, seenEdges, statementId, fileId, relation);
    }

    private static IEnumerable<string> PerformTargets(string raw)
    {
        var tokens = SplitTokens(raw).ToArray();
        if (tokens.Length < 2 || new[] { "UNTIL", "VARYING", "TIMES" }.Contains(tokens[1], StringComparer.OrdinalIgnoreCase))
        {
            yield break;
        }

        yield return CleanToken(tokens[1]);
        for (var i = 2; i < tokens.Length - 1; i++)
        {
            if (tokens[i].Equals("THRU", StringComparison.OrdinalIgnoreCase) || tokens[i].Equals("THROUGH", StringComparison.OrdinalIgnoreCase))
            {
                yield return CleanToken(tokens[i + 1]);
            }
        }
    }

    private static IEnumerable<string> ReferencedVariables(string raw, IReadOnlySet<string> knownVariables)
    {
        foreach (var variableName in knownVariables.OrderByDescending(v => v.Length))
        {
            if (Regex.IsMatch(raw, $"(?<![A-Z0-9-]){Regex.Escape(variableName)}(?![A-Z0-9-])", RegexOptions.IgnoreCase))
            {
                yield return variableName;
            }
        }
    }

    private static void AddNode(GraphifyGraph graph, HashSet<string> seenNodes, string id, string label, string type, Dictionary<string, object?> attributes)
    {
        if (!seenNodes.Add(id))
        {
            return;
        }

        var node = new Dictionary<string, object?> { ["id"] = id, ["label"] = label, ["type"] = type };
        foreach (var pair in attributes.Where(pair => pair.Value is not null))
        {
            node[pair.Key] = pair.Value;
        }

        graph.Nodes.Add(node);
    }

    private static void AddEdge(GraphifyGraph graph, HashSet<string> seenEdges, string source, string target, string relation)
    {
        var key = $"{source}|{target}|{relation}";
        if (seenEdges.Add(key))
        {
            graph.Edges.Add(new Dictionary<string, object?> { ["source"] = source, ["target"] = target, ["relation"] = relation });
        }
    }

    private static string Id(string kind, string label)
    {
        var value = Regex.Replace(CleanToken(label).ToLowerInvariant(), "[^a-z0-9_-]+", "-").Trim('-');
        return $"{kind}:{(string.IsNullOrWhiteSpace(value) ? "unknown" : value)}";
    }

    private static string CleanToken(string value) => value.Trim().TrimEnd('.').Trim('"', '\'');

    private static string Short(string value) => value.Length <= 120 ? value : value[..119] + "…";

    private static IReadOnlyList<string> SplitTokens(string raw) => raw.Trim().TrimEnd('.').Split(' ', StringSplitOptions.RemoveEmptyEntries);
}

/// <summary>Graphify-compatible graph data transfer object.</summary>
public sealed class GraphifyGraph
{
    /// <summary>Graph nodes.</summary>
    public List<Dictionary<string, object?>> Nodes { get; } = [];

    /// <summary>Graph edges.</summary>
    public List<Dictionary<string, object?>> Edges { get; } = [];
}

File: src/export/graphify_exporter.py
from __future__ import annotations

import json
import re
from pathlib import Path
from typing import Any, Iterable

from src.program.programm import Programm
from src.statements import Statement
from src.statements.zcallpgm_statement import ZCallPgmStatement
from src.var import Variable

Graph = dict[str, list[dict[str, Any]]]


def export_program_to_graphify(program: Programm, *, source_file: str | None = None) -> Graph:
    """Return a Graphify-compatible graph for a parsed COBOL program."""
    graph: Graph = {"nodes": [], "edges": []}
    seen_nodes: set[str] = set()
    seen_edges: set[tuple[str, str, str]] = set()

    program_label = Path(source_file or program.name).stem.upper()
    program_id = _id("program", program_label)
    _node(graph, seen_nodes, program_id, label=program_label, type="program", source_file=source_file or program.name)

    for variable in getattr(program.memory_stack, "stack", []) or []:
        variable_id = _export_variable(graph, seen_nodes, seen_edges, variable)
        _edge(graph, seen_edges, program_id, variable_id, "declares")

    known_variables = set(getattr(program.memory_stack, "known_variables", []) or [])

    for paragraph_index, paragraph in enumerate(program.paragraphs):
        paragraph_label = _clean(getattr(paragraph, "statement", f"paragraph-{paragraph_index}"))
        paragraph_id = _id("paragraph", f"{program_label}:{paragraph_label}:{paragraph_index}")
        _node(graph, seen_nodes, paragraph_id, label=paragraph_label, type="paragraph", source_file=source_file or program.name)
        _edge(graph, seen_edges, program_id, paragraph_id, "contains")

        for statement_index, statement in enumerate(_children(paragraph)):
            _export_statement(
                graph,
                seen_nodes,
                seen_edges,
                paragraph_id,
                statement,
                statement_index,
                known_variables,
                source_file or program.name,
            )

    return graph


def write_graphify_export(program: Programm, output_path: str | Path, *, source_file: str | None = None) -> Path:
    """Write the Graphify-compatible export as JSON."""
    path = Path(output_path)
    path.parent.mkdir(parents=True, exist_ok=True)
    graph = export_program_to_graphify(program, source_file=source_file)
    path.write_text(json.dumps(graph, indent=2, ensure_ascii=False), encoding="utf-8")
    return path


def _export_variable(graph: Graph, seen_nodes: set[str], seen_edges: set[tuple[str, str, str]], variable: Variable) -> str:
    variable_name = variable.name or variable.statement
    variable_id = _id("variable", variable_name)
    _node(
        graph,
        seen_nodes,
        variable_id,
        label=variable_name,
        type="variable",
        level=variable.number,
        pic=variable.pic,
        usage=variable.usage,
        redefines=variable.redefines,
        occurs=variable.occurs,
        occurs_min=variable.occurs_min,
        occurs_max=variable.occurs_max,
        depends_on=variable.depends_on,
        is_group=variable.is_group,
    )

    if variable.redefines:
        target_id = _id("variable", variable.redefines)
        _node(graph, seen_nodes, target_id, label=variable.redefines, type="variable", inferred=True)
        _edge(graph, seen_edges, variable_id, target_id, "redefines")

    if variable.depends_on:
        target_id = _id("variable", variable.depends_on)
        _node(graph, seen_nodes, target_id, label=variable.depends_on, type="variable", inferred=True)
        _edge(graph, seen_edges, variable_id, target_id, "occurs_depends_on")

    for definition in variable.definitions:
        child_id = _export_variable(graph, seen_nodes, seen_edges, definition)
        _edge(graph, seen_edges, variable_id, child_id, "defines_condition")

    for child in variable.variables:
        child_id = _export_variable(graph, seen_nodes, seen_edges, child)
        _edge(graph, seen_edges, variable_id, child_id, "contains")

    return variable_id


def _export_statement(
    graph: Graph,
    seen_nodes: set[str],
    seen_edges: set[tuple[str, str, str]],
    parent_id: str,
    statement: Statement,
    statement_index: int,
    known_variables: set[str],
    source_file: str,
) -> str:
    raw = getattr(statement, "statement", "") or getattr(statement, "raw_text", "") or str(statement)
    statement_id = _id("statement", f"{parent_id}:{statement_index}:{raw}")
    statement_type = getattr(statement, "typeStatement", None) or statement.__class__.__name__.replace("Statement", "").lower()

    _node(graph, seen_nodes, statement_id, label=_short(raw), type="statement", statement_type=statement_type, source_file=source_file)
    _edge(graph, seen_edges, parent_id, statement_id, "contains")
    _semantic_edges(graph, seen_nodes, seen_edges, statement_id, statement, raw)

    for variable_name in _referenced_variables(raw, known_variables):
        variable_id = _id("variable", variable_name)
        _node(graph, seen_nodes, variable_id, label=variable_name, type="variable", inferred=True)
        _edge(graph, seen_edges, statement_id, variable_id, "references")

    for child_index, child in enumerate(_children(statement)):
        _export_statement(graph, seen_nodes, seen_edges, statement_id, child, child_index, known_variables, source_file)

    return statement_id


def _semantic_edges(graph: Graph, seen_nodes: set[str], seen_edges: set[tuple[str, str, str]], statement_id: str, statement: Statement, raw: str) -> None:
    upper = raw.strip().upper()

    if isinstance(statement, ZCallPgmStatement) or _is_zcallpgm(raw):
        target = getattr(statement, "target_pgm", None) or _zcallpgm_target(raw)
        if target:
            target_id = _id("program", target)
            _node(graph, seen_nodes, target_id, label=target, type="program", inferred=True)
            _edge(graph, seen_edges, statement_id, target_id, "calls")
        for copy_name in getattr(statement, "copies", []) or []:
            copy_id = _id("copy", copy_name)
            _node(graph, seen_nodes, copy_id, label=copy_name, type="copy", inferred=True)
            _edge(graph, seen_edges, statement_id, copy_id, "uses_copy_argument")
        return

    if upper.startswith("CALL "):
        target = _call_target(raw)
        if target:
            target_id = _id("program", target)
            _node(graph, seen_nodes, target_id, label=target, type="program", inferred=True)
            _edge(graph, seen_edges, statement_id, target_id, "calls")
        return

    if upper.startswith("PERFORM "):
        for target in _perform_targets(raw):
            target_id = _id("paragraph", target)
            _node(graph, seen_nodes, target_id, label=target, type="paragraph", inferred=True)
            _edge(graph, seen_edges, statement_id, target_id, "performs")
        return

    if upper.startswith("READ "):
        target = _first_operand(raw)
        if target:
            file_id = _id("file", target)
            _node(graph, seen_nodes, file_id, label=target, type="file", inferred=True)
            _edge(graph, seen_edges, statement_id, file_id, "reads_from")

    if upper.startswith("WRITE ") or upper.startswith("REWRITE "):
        target = _first_operand(raw)
        if target:
            file_id = _id("file", target)
            _node(graph, seen_nodes, file_id, label=target, type="file", inferred=True)
            _edge(graph, seen_edges, statement_id, file_id, "writes_to")


def _children(statement: Statement) -> Iterable[Statement]:
    for child in getattr(statement, "statements", []) or []:
        yield child
    for child in getattr(statement, "then_statements", []) or []:
        yield child
    for child in getattr(statement, "else_statements", []) or []:
        yield child
    for branch in getattr(statement, "whens", []) or []:
        for child in getattr(branch, "statements", []) or []:
            yield child
    for child in getattr(statement, "when_other", []) or []:
        yield child


def _id(kind: str, label: str) -> str:
    value = _clean(str(label)).lower()
    value = re.sub(r"[^a-z0-9_-]+", "-", value).strip("-") or "unknown"
    return f"{kind}:{value}"


def _node(graph: Graph, seen: set[str], node_id: str, **attrs: Any) -> None:
    if node_id in seen:
        return
    seen.add(node_id)
    graph["nodes"].append({"id": node_id, **{key: value for key, value in attrs.items() if value is not None}})


def _edge(graph: Graph, seen: set[tuple[str, str, str]], source: str, target: str, relation: str) -> None:
    key = (source, target, relation)
    if key in seen:
        return
    seen.add(key)
    graph["edges"].append({"source": source, "target": target, "relation": relation})


def _clean(value: str) -> str:
    return value.strip().rstrip(".")


def _short(value: str, max_length: int = 120) -> str:
    value = value.strip()
    return value if len(value) <= max_length else value[: max_length - 1] + "…"


def _is_zcallpgm(raw: str) -> bool:
    tokens = raw.strip().split()
    return len(tokens) >= 3 and tokens[0].upper() == "CALL" and tokens[1].strip("'\"").upper() == "ZCALLPGM"


def _zcallpgm_target(raw: str) -> str | None:
    tokens = raw.strip().split()
    if len(tokens) >= 4 and tokens[2].upper() == "USING":
        return tokens[3].strip("'\"")
    return None


def _call_target(raw: str) -> str | None:
    tokens = raw.strip().split()
    return tokens[1].strip("'\"") if len(tokens) >= 2 else None


def _perform_targets(raw: str) -> list[str]:
    tokens = raw.strip().rstrip(".").split()
    if len(tokens) < 2 or tokens[1].upper() in {"UNTIL", "VARYING", "TIMES"}:
        return []
    targets = [tokens[1]]
    for keyword in ("THRU", "THROUGH"):
        for index, token in enumerate(tokens):
            if token.upper() == keyword and index + 1 < len(tokens):
                targets.append(tokens[index + 1])
    return [_clean(target) for target in targets]


def _first_operand(raw: str) -> str | None:
    tokens = raw.strip().rstrip(".").split()
    return tokens[1] if len(tokens) >= 2 else None


def _referenced_variables(raw: str, variable_names: set[str]) -> list[str]:
    upper_raw = raw.upper()
    found: list[str] = []
    for variable_name in sorted(variable_names, key=len, reverse=True):
        pattern = rf"(?<![A-Z0-9-]){re.escape(variable_name.upper())}(?![A-Z0-9-])"
        if re.search(pattern, upper_raw):
            found.append(variable_name)
    return found
